api/engineapi/routes/contracts.py

```diff
@@ -0,0 +1,203 @@
+"""
+Contract registration API
+
+Moonstream users can register contracts on Moonstream Engine. This allows them to use these contracts
+as part of their chain-adjacent activities (like performing signature-based token distributions on the
+Dropper contract).
+"""
+import logging
+from typing import Dict, List, Optional
+from uuid import UUID
+
+from fastapi import Body, Depends, FastAPI, Query, Request
+from fastapi.middleware.cors import CORSMiddleware
+from sqlalchemy.orm import Session
+
+from .. import data, db
+from .. import contracts_actions
+from ..middleware import BroodAuthMiddleware, EngineHTTPException
+from ..settings import DOCS_TARGET_PATH, ORIGINS
+from ..version import VERSION
+
+logger = logging.getLogger(__name__)
+
+
+TITLE = "Moonstream Engine Contracts API"
+DESCRIPTION = "Users can register contracts on the Moonstream Engine for use in chain-adjacent activities, like setting up signature-based token distributions."
+
+
+tags_metadata = [
+    {
+        "name": "contracts",
+        "description": DESCRIPTION,
+    }
+]
+
+
+whitelist_paths = {
+    "/contracts/openapi.json": "GET",
+    f"/contracts/{DOCS_TARGET_PATH}": "GET",
+    "/contracts/types": "GET",
+    "/contracts/requests": "GET",
+}
+
+app = FastAPI(
+    title=TITLE,
+    description=DESCRIPTION,
+    version=VERSION,
+    openapi_tags=tags_metadata,
+    openapi_url="/openapi.json",
+    docs_url=None,
+    redoc_url=f"/{DOCS_TARGET_PATH}",
+)
+
+
+app.add_middleware(BroodAuthMiddleware, whitelist=whitelist_paths)
+
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=ORIGINS,
+    allow_credentials=True,
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+
+@app.get("/types")
+async def contract_types() -> Dict[str, str]:
+    """
+    Describes the contract_types that users can register contracts as against this API.
+    """
+    return {
+        contracts_actions.ContractType.raw.value: "A generic smart contract. You can ask users to submit arbitrary calldata to this contract.",
+        contracts_actions.ContractType.dropper.value: "A Dropper contract. You can authorize users to submit claims against this contract.",
+    }
+
+
+@app.get("/")
+async def list_registered_contracts(
+    request: Request,
+    blockchain: Optional[str] = Query(None),
+    address: Optional[str] = Query(None),
+    contract_type: Optional[contracts_actions.ContractType] = Query(None),
+    limit: int = Query(10),
+    offset: Optional[int] = Query(None),
+    db_session: Session = Depends(db.yield_db_session),
```

Comment:
This is read handler, better to use read only database connection 
`db.yield_db_read_only_session`